domains/runtime/evm/src/lib.rs

```diff
@@ -837,19 +916,34 @@ fn check_transaction_and_do_pre_dispatch_inner(
     // which would help to maintain context across multiple transaction validity check against same
     // runtime instance.
     match xt.signed {
-        CheckedSignature::Signed(account_id, extra) => extra
-            .pre_dispatch(&account_id, &xt.function, &dispatch_info, encoded_len)
-            .map(|_| ()),
+        CheckedSignature::Signed(account_id, extra) => {
+            // if a sender sends a one evm transaction first and substrate transaction
+            // after with same nonce, then reject the second transaction
+            // if sender reverse the transaction types, substrate first and evm second,
+            // since substrate updates nonce in pre_dispatch, evm transaction will be rejected.
+            if let Some(tracked_nonce) = EVMNoncetracker::account_nonce(H160::from(account_id.0)) {
+                let account_nonce = U256::from(System::account_nonce(account_id));
+                let current_nonce = max(tracked_nonce, account_nonce);
+                let transaction_nonce = U256::from(extra.5 .0);
+                match transaction_nonce.cmp(&current_nonce) {
+                    Ordering::Less => return Err(InvalidTransaction::Stale.into()),
+                    Ordering::Greater => return Err(InvalidTransaction::Future.into()),
+                    Ordering::Equal => {}
+                }
+            }
+
+            extra
+                .pre_dispatch(&account_id, &xt.function, &dispatch_info, encoded_len)
+                .map(|_| ())
```

Comment:
It seems there is still issue, the evm tx and substrate tx are essentially updating nonce in 2 different places, if there is an evm tx with nonce 1 then a substrate tx with nonce 2, the substrate tx will fail in the substrate pre_dispatch since there is also nonce check and substrate nonce is not updated during the evm tx so it is expecting nonce 1.

Reply:
> It seems there is still issue, the evm tx and substrate tx are essentially updating nonce in 2 different places

No, this maybe incorrect assumption. `System::Account::nonce` will be updated post EVM execution and substrate already updates it. So nonce location is always the same. 
Maybe EVM does also updates the nonce in other places but it still updates here. So this should not be a problem

Reply:
What I referring to is the `pre_dispatch` during bundle construction not block execution. In `check_extrinsics_and_do_pre_dispatch`, the evm tx will update nonce in `EVMNoncetracker`, while the substrate will also check against `EVMNoncetracker` but then will fail in the substrate `pre_dispatch` in L936 so it will not include in the bundle.

Reply:
I see good catch!
I guess we need to define a custom SignedExtra for substrate as well. This is quiet unfortunate :(